File: python/problems/leetcode/0023_merge_k_sorted_lists/solution.py
"""LeetCode 23: Merge k Sorted Lists."""

from __future__ import annotations

import heapq
from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: list[Optional[ListNode]]) -> Optional[ListNode]:
        heap: list[tuple[int, int, ListNode]] = []
        for i, head in enumerate(lists):
            if head is not None:
                heapq.heappush(heap, (head.val, i, head))
        dummy = ListNode()
        cur = dummy
        counter = len(lists)
        while heap:
            _, _, node = heapq.heappop(heap)
            cur.next = node
            cur = node
            if node.next is not None:
                heapq.heappush(heap, (node.next.val, counter, node.next))
                counter += 1
        cur.next = None
        return dummy.next


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    lists = [from_list([1, 4, 5]), from_list([1, 3, 4]), from_list([2, 6])]
    assert to_list(Solution().mergeKLists(lists)) == [1, 1, 2, 3, 4, 4, 5, 6]
    assert Solution().mergeKLists([]) is None
    assert to_list(Solution().mergeKLists([None, from_list([1])])) == [1]
    print("0023 OK")

File: cpp/problems/leetcode/0023_merge_k_sorted_lists/solution.cpp
// LeetCode 23
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* n) : val(x), next(n) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) { return a->val > b->val; };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);
        for (auto* h : lists)
            if (h) pq.push(h);
        ListNode dummy;
        ListNode* cur = &dummy;
        while (!pq.empty()) {
            ListNode* top = pq.top();
            pq.pop();
            cur->next = top;
            cur = top;
            if (top->next) pq.push(top->next);
        }
        cur->next = nullptr;
        return dummy.next;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    vector<ListNode*> lists{fromList({1, 4, 5}), fromList({1, 3, 4}), fromList({2, 6})};
    assert((toList(Solution().mergeKLists(lists)) ==
            vector<int>{1, 1, 2, 3, 4, 4, 5, 6}));
    vector<ListNode*> empty{};
    assert(Solution().mergeKLists(empty) == nullptr);
    vector<ListNode*> partial{nullptr, fromList({1})};
    assert((toList(Solution().mergeKLists(partial)) == vector<int>{1}));
    cout << "0023 OK" << endl;
    return 0;
}
